tests/binary_agreement.rs

```diff
@@ -1,108 +1,137 @@
-    }
+/// Test configuration for Binary Agreement tests.
+#[derive(Debug)]
+struct TestConfig {
+    /// The desired network dimension.
+    dimension: NetworkDimension,
+    /// Random number generator to be passed to subsystems.
+    seed: TestRngSeed,
+}
 
-    // Handle messages in random order until all nodes have output the proposed value.
-    while !network.nodes.values().all(TestNode::terminated) {
-        network.step();
+prop_compose! {
+    /// Strategy to generate a test configuration.
+    fn arb_config()
+        (
+            dimension in NetworkDimension::range(1, 50),
+            seed in gen_seed()
+        ) -> TestConfig
+    {
+        TestConfig { dimension, seed }
     }
-    // Verify that all instances output the same value.
-    let mut expected = input;
-    for node in network.nodes.values() {
-        if let Some(b) = expected {
-            assert!(once(&b).eq(node.outputs()));
-        } else {
-            assert_eq!(1, node.outputs().len());
-            expected = Some(node.outputs()[0]);
-        }
+}
+
+/// Proptest wrapper for `binary_agreement`.
+proptest!{
+    #![proptest_config(ProptestConfig {
+        cases: 1, .. ProptestConfig::default()
+    })]
+    #[test]
+    #[cfg_attr(feature = "cargo-clippy", allow(unnecessary_operation))]
+    fn binary_agreement_wrapper(cfg in arb_config()) {
+        binary_agreement(cfg)
     }
-    assert!(expected.iter().eq(network.observer.outputs()));
 }
 
-fn test_binary_agreement_different_sizes<A, F>(new_adversary: F)
-where
-    A: Adversary<BinaryAgreement<NodeId, u8>>,
-    F: Fn(usize, usize) -> A,
-{
-    // This returns an error in all but the first test.
-    let _ = env_logger::try_init();
+type NodeId = u16;
+type Algo = BinaryAgreement<NodeId, u8>;
 
-    let mut rng = rand::thread_rng();
-    let sizes = (1..6)
-        .chain(once(rng.gen_range(6, 20)))
-        .chain(once(rng.gen_range(30, 50)));
-    for size in sizes {
-        let num_faulty_nodes = (size - 1) / 3;
-        let num_good_nodes = size - num_faulty_nodes;
-        for &input in &[None, Some(false), Some(true)] {
-            info!(
-                "Test start: {} good nodes and {} faulty nodes, input: {:?}",
-                num_good_nodes, num_faulty_nodes, input
-            );
-            let adversary = |_| new_adversary(num_good_nodes, num_faulty_nodes);
-            let new_ba = |netinfo: Arc<NetworkInfo<NodeId>>| {
-                BinaryAgreement::new(netinfo, 0).expect("Binary Agreement instance")
-            };
-            let network = TestNetwork::new(num_good_nodes, num_faulty_nodes, adversary, new_ba);
-            test_binary_agreement(network, input);
-            info!(
-                "Test success: {} good nodes and {} faulty nodes, input: {:?}",
-                num_good_nodes, num_faulty_nodes, input
-            );
+impl VirtualNet<Algo> {
+    fn test_binary_agreement<R>(&mut self, input: Option<bool>, mut rng: R)
```

Comment:
Should that really be a method of `VirtualNet`, instead of a separate function? (@mbr?)

Reply:
As a private method, it makes a perfect sense to me.

Reply:
I'm skeptical as well. I don't have a great retort at this point though =).